src/api/spec/controllers/webui/session_controller_spec.rb

```diff
@@ -54,4 +55,71 @@
       it_behaves_like 'login'
     end
   end
+
+  context 'in kerberos mode' do
+    before do
+      stub_const('CONFIG', CONFIG.merge('kerberos_service_principal' => 'HTTP/obs.test.com@test_realm.com',
+                                        'kerberos_realm'             => 'test_realm.com',
+                                        'kerberos_mode'              => true))
+    end
+
+    context 'for a request that requires authentication' do
+      render_views
+      before do
+        get :new
+      end
+
+      context "and 'Negotiate' header is not set" do
+        it 'informs the client tool (browser) that kerberos authentication is required' do
+          expect(response.headers['WWW-Authenticate']).to eq('Negotiate')
+          expect(response.status).to eq(401)
+        end
+
+        it 'informs users about failed kerberos authentication and possible cause' do
+          expect(response.body).to have_text('Kerberos authentication required')
+          expect(response.body).to have_text('You are seeing this page, because you are ' \
+                                    "not authenticated in the kerberos realm ('test_realm.com').")
+        end
+      end
+    end
+
+    context 'for a request with valid kerberos ticket' do
+      include_context 'a kerberos mock for' do
+        let(:login) { user.login }
+        let(:ticket) { 'krb5_ticket' }
+      end
+
+      render_views
+      it 'authenticates the user' do
+        # In real life done by the browser / client
+        request.headers['AUTHORIZATION'] = "Negotiate #{Base64.strict_encode64(ticket)}"
+
+        get :new
+        expect(response).to redirect_to root_path
+        expect(User.current).to eq(user)
+        expect(session[:login]).to eq(user.login)
+      end
+    end
+
+    context 'for a request where GSSAPI raises an exception' do
+      let(:gssapi_mock) { double(:gssapi) }
+
+      before do
+        allow(gssapi_mock).to receive(:acquire_credentials).
+          and_raise(GSSAPI::GssApiError, "couldn't validate ticket")
+
+        allow(GSSAPI::Simple).to receive(:new).with(
+          'obs.test.com', 'HTTP', '/etc/krb5.keytab'
+        ).and_return(gssapi_mock)
+      end
+
+      it 'does not authenticate the user' do
+        request.headers['AUTHORIZATION'] = "Negotiate #{Base64.strict_encode64('ticket')}"
+
+        get :new
+        expect(response).to redirect_to root_path
+        expect(flash[:error]).to eq("Authentication failed: 'Received a GSSAPI exception; couldn't validate ticket: couldn't validate ticket.'")
```

Comment:
We had some discussion in the pass about testing the string in the errors as this give as little flexibility, making the test failing if the error message is slightly modified.

Reply:
Actually I like that it's verifying the error is bullshit :)

I mean, the error it couldn't validate ticket is that it couldn't validate ticket

Reply:
I still think it makes sense to check what exact error happened. And I can't remember when was the last time I had to change the string of an error message in a test:-)

Reply:
I agree, it's quite easy to mess up error messages during refactoring - and it's an important part of the UI.

Reply:
The message itself comes from the [GSSAPI gem](https://github.com/openSUSE/open-build-service/blob/master/src/api/lib/authenticator.rb#L157)

Reply:
IMO it's not worth the hassle

Reply:
Usually comparing error messages in a spec is a symptom of not testing what you really expect to happen.
In this case what you expect not to happen is that `User.current` is set for instance.

Reply:
well, I expect the error message to be valueable to users. App developers tend to put a "an error happened" in there, that is not my expectation.

Reply:
but could we please concentrate on the topic: I just moved the expectations to a different spec kind

Reply:
> I expect the error message to be valueable to users

We don't test for developer stupidity. We test for correct functionality.

> I just moved the expectations to a different spec kind

Sure...